### PR TITLE
ARROW-17989: [C++][Python] Enable struct_field kernel to accept string field names

### PR DESCRIPTION
**Pull request by milesgranger**
Will close [ARROW-17989](https://issues.apache.org/jira/browse/ARROW-17989)

Allows using names in `pc.struct_field`
```python
In [1]: arr = pa.array([{'a': {'b': 1}, 'c': 2}])

In [2]: pc.struct_field(arr, 'c')
Out[2]:
<pyarrow.lib.Int64Array object at 0x7f1442da3d60>
[
  2
]

In [3]: pc.struct_field(arr, '.a.b')
Out[3]:
<pyarrow.lib.Int64Array object at 0x7f14436d0f40>
[
  1
]

# And indices as before...
In [4]: pc.struct_field(arr, [0, 0])
Out[4]:
<pyarrow.lib.Int64Array object at 0x7f14436d0ee0>
[
  1
]

In [5]:
```

### CODE REVIEW

File: cpp/src/arrow/compute/api_scalar.cc [hunk 1]
```diff
@@ -560,8 +560,12 @@ StrptimeOptions::StrptimeOptions(std::string format, TimeUnit::type unit,
 StrptimeOptions::StrptimeOptions() : StrptimeOptions("", TimeUnit::MICRO, false) {}
 constexpr char StrptimeOptions::kTypeName[];
 
+StructFieldOptions::StructFieldOptions(FieldRef field_ref)
+    : FunctionOptions(internal::kStructFieldOptionsType), field_ref(field_ref) {}
 StructFieldOptions::StructFieldOptions(std::vector<int> indices)
-    : FunctionOptions(internal::kStructFieldOptionsType), indices(std::move(indices)) {}
+    : StructFieldOptions(FieldRef(FieldPath(std::move(indices)))) {}
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/api_scalar.cc:566]: Should we set both FieldRef and indices here? (for backwards compatibility of the indices field in the options class)

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 2]
```diff
@@ -252,6 +266,35 @@ struct StructFieldFunctor {
...
+      current = array.GetFieldByName(*field_ref.name());
+      if (current == nullptr) {
+        return Status::Invalid("Field not found in struct: '", *field_ref.name(), "'");
+      }
+    } else if (field_ref.IsFieldPath()) {
+      for (const auto& idx : field_ref.field_path()->indices()) {
+        ARROW_RETURN_NOT_OK(CheckIndex(idx, *current->type()));
+        const auto& array = checked_cast<const StructArray&>(*current);
+        ARROW_ASSIGN_OR_RAISE(current, array.GetFlattenedField(idx, ctx->memory_pool()));
+      }
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:288]: There is a `FieldPath::Get()` for arrays, could that be used here instead of this loop?

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 3]
```diff
@@ -252,6 +266,35 @@ struct StructFieldFunctor {
...
+    } else {
+      DCHECK(field_ref.IsNested());
+      for (const auto& ref : *field_ref.nested_refs()) {
+        ARROW_ASSIGN_OR_RAISE(current, ApplyFieldRef(ctx, ref, current));
+      }
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:293]: Possible alternative: convert the nested field ref to a FieldPath (eg with FieldRef::FindOne?) before the code to handle a FieldPath, and then it can use the same code.

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 4]
```diff
@@ -269,14 +312,34 @@ struct StructFieldFunctor {
   }
 };
 
+Result<const DataType*> RecursiveResolveStructFieldType(const FieldRef& field_ref,
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:315]: This could maybe also be replaced or simplified with one of the existing FieldRef / FieldPath utilities?

File: python/pyarrow/tests/test_compute.py [hunk 5]
```diff
@@ -2691,16 +2691,28 @@ def test_struct_fields_options():
...
-
-    assert pc.struct_field(arr,
-                           indices=[1, 1]) == pa.array(["bar", None, ""])
-    assert pc.struct_field(arr, [1, 1]) == pa.array(["bar", None, ""])
-    assert pc.struct_field(arr, [0]) == pa.array([4, 5, 6], type=pa.int64())
+    assert pc.struct_field(arr, '.c.b', is_dot_path=True) == b
+    assert pc.struct_field(arr, '.a', is_dot_path=True) == a
+    assert pc.struct_field(arr, 'a') == a
+    assert pc.struct_field(arr, indices=[1, 1]) == b
+    assert pc.struct_field(arr, [1, 1]) == b
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_compute.py:2698]: I think it should also be possible to pass the names as a list? (`['b', 'c']`)

**Review by jorisvandenbossche — COMMENTED** [review]
Nice! Some quick questions / suggestions

File: cpp/src/arrow/compute/kernels/scalar_nested.cc (hunk 2 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/compute/kernels/scalar_nested.cc:288]: Hmm, but the GetFlattenedField ensures that null bitmaps are combined, which is maybe not the case for FieldPath::Get() (since that doesn't use GetFlattenedField, it's unlikely that it takes that into account)

(it would also be good to have tests for this situation, in case that wouldn't yet be the case)

File: python/pyarrow/tests/test_compute.py (hunk 5 above)
Comment (milesgranger) [reply python/pyarrow/tests/test_compute.py:2698]: Which would be equivalent to a dot path of `.b.c`, is that right?

Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_compute.py:2698]: Yes, and we should probably allow passing an actual FieldRef instance as well (`pc.field(("b", "c"))`)

Personally, I am not sure we necessarily need to support a dotted path here, especially if it requires a keyword (which seems to defeat the purpose of added convenience a bit?)

Comment (milesgranger) [reply python/pyarrow/tests/test_compute.py:2698]: Could do the same assumption `FromDotPath` does and create one if it has a leading `.`, then remove the kwarg?

Comment (milesgranger) [reply python/pyarrow/tests/test_compute.py:2707]: Done in https://github.com/apache/arrow/pull/14495/commits/0ff256726c100e791378534c9a7fdbd8d5c6dc1b and https://github.com/apache/arrow/pull/14495/commits/a4b5ce00dae66fe9a329359700d3db767eb31e20

File: cpp/src/arrow/compute/kernels/scalar_nested.cc (hunk 2 above)
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested.cc]: I gave it a go. Seems those expected null checks already exist as C++ tests started failing with:
`Null counts differ. Expected -1 but was 0` and similar discrepancies.

File: cpp/src/arrow/compute/kernels/scalar_nested.cc (hunk 4 above)
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested.cc]: Good eye! :eyes:

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 6]
```diff
@@ -252,6 +266,39 @@ struct StructFieldFunctor {
...
+                                                      const FieldRef& field_ref,
+                                                      std::shared_ptr<Array> current) {
+    if (current->type_id() != Type::STRUCT) {
+      return Status::Invalid("Not a StructArray: ", current->ToString(),
+                             "\nMaybe a bad FieldRef? ", field_ref.ToString());
+    }
+
+    if (field_ref.IsName()) {
+      const auto& array = checked_cast<const StructArray&>(*current);
+      current = array.GetFieldByName(*field_ref.name());
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:279]: I think `GetFieldByName` actually has the same concern as mentioned before about not taking into account top-level nulls? (so also for this case can convert to FieldPath with FindOne, and then this can be simplified a bit further) 
(if that's the case, that probably indicates the tests needs to be expanded with a case of using a named FieldRef in the test with nulls)

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 7]
```diff
@@ -271,11 +318,30 @@ struct StructFieldFunctor {
...
-    RETURN_NOT_OK(StructFieldFunctor::CheckIndex(index, *type));
-    type = type->field(index)->type().get();
+
+  if (field_ref.IsName()) {
+    for (const auto& path : field_ref.FindAll(*type)) {
+      for (const auto& index : path.indices()) {
+        RETURN_NOT_OK(StructFieldFunctor::CheckIndex(index, *type));
+        type = type->field(index)->type().get();
+      }
+    }
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:330]: The fact that you use FindAll here, that would catch duplicate field names? Maybe we don't necessarily have to support that? (also the way that you loop over those might not be correct?)

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 8]
```diff
@@ -271,11 +318,30 @@ struct StructFieldFunctor {
...
+  } else {
+    DCHECK(field_ref.IsFieldPath() || field_ref.IsNested());
+
+    FieldPath field_path;
+    if (field_ref.IsNested()) {
+      ARROW_ASSIGN_OR_RAISE(field_path, field_ref.FindOne(*type));
+    } else {
+      field_path = *field_ref.field_path();
+    }
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:339]: I am wondering if it is actually needed to do this if/else. Also a FieldRef backed by a field path has a FindOne method, and in that case it will basically return the same field path, while still checking that it is a valid field path (not out of bounds)

**Review by jorisvandenbossche — COMMENTED** [review]
A few more comments!

Can you also add a basic C++ test for the case of passing a FieldRef to struct_field?

File: cpp/src/arrow/compute/kernels/scalar_nested.cc (hunk 8 above)
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested.cc]: Seems like it is needed, removing if/else and using `field_ref.FindOne(*type)` will abort when `pc.struct_field(arr, [])` (empty indices) with:
```
pyarrow/tests/test_compute.py::test_struct_fields_options /home/milesg/Projects/arrow/cpp/src/arrow/type.cc:1360:  Check failed: _s.ok() Operation failed: maybe_field.status()
Bad status: Invalid: empty indices cannot be traversed
```
Seems like another if statement would be needed to handle that case, so maybe stick with this?

Comment (jorisvandenbossche) [reply cpp/src/arrow/compute/kernels/scalar_nested.cc]: I am wondering if it is actually intentional that we support empty index (I don't see it covered in the C++ tests?). That doesn't sound very useful? (it just returns the original array?)

Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested.cc]: The empty index case was in the Python tests. 

I suppose it may be useful if only to serve the purpose of idempotent behavior, similar to `[:]` or `start==stop` -> empty lists in the `list_slice` kernel. Albeit those are unique to slicing.

File: cpp/src/arrow/compute/api_scalar.h [hunk 9]
```diff
@@ -277,12 +277,16 @@ class ARROW_EXPORT SetLookupOptions : public FunctionOptions {
...
+  explicit StructFieldOptions(std::initializer_list<int>);
+  explicit StructFieldOptions(FieldRef field_ref,
+                              std::vector<int> indices = std::vector<int>());
   StructFieldOptions();
   static constexpr char const kTypeName[] = "StructFieldOptions";
 
   /// The child indices to extract. For instance, to get the 2nd child
   /// of the 1st child of a struct or union, this would be {0, 1}.
   std::vector<int> indices;
+  FieldRef field_ref;
```
Comment (pitrou) [inline cpp/src/arrow/compute/api_scalar.h:289]: You probably don't need to keep the `std::vector<int> indices` since the `FieldRef` is able to replace them?

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 10]
```diff
@@ -252,6 +266,29 @@ struct StructFieldFunctor {
     return Status::OK();
   }
 
+  static Result<std::shared_ptr<Array>> ApplyFieldRef(KernelContext* ctx,
+                                                      const FieldRef& field_ref,
+                                                      std::shared_ptr<Array> current) {
+    if (current->type_id() != Type::STRUCT) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:272]: Hmm, why wouldn't this function work on unions?

File: cpp/src/arrow/compute/kernels/scalar_nested.cc [hunk 11]
```diff
@@ -196,9 +196,22 @@ const FunctionDoc list_element_doc(
 struct StructFieldFunctor {
   static Status Exec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
     const auto& options = OptionsWrapper<StructFieldOptions>::Get(ctx);
-
     std::shared_ptr<Array> current = MakeArray(batch[0].array.ToArrayData());
-    for (const auto& index : options.indices) {
+
+    // Specific struct handling
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_nested.cc:201]: I don't understand why this is special-cased. I would intuitively expect the following process:
1) compute a `FieldPath` using `FieldRef::FindOne`
2) Loop over the indices of the `FieldPath` in the loop below

File: cpp/src/arrow/compute/api_scalar.h (hunk 9 above)
Comment (milesgranger) [reply cpp/src/arrow/compute/api_scalar.h:289]: In the JIRA issue, [seems it should be kept](https://issues.apache.org/jira/browse/ARROW-17989?focusedCommentId=17617009&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17617009). I tried `ARROW_DEPRECATED` above the public `indices` member which it didn't like since that's one of the places we declare it in the first place. 


> /home/milesg/Projects/arrow/cpp/src/arrow/compute/api_scalar.h:277:20: error: 'arrow::compute::StructFieldOptions::indices' is deprecated: Deprecated since 11.0.0. Use `field_ref` instead. [-Werror=deprecated-declarations]
  277 | class ARROW_EXPORT StructFieldOptions : public FunctionOptions {
      |                    ^~~~~~~~~~~~~~~~~~
/home/milesg/Projects/arrow/cpp/src/arrow/compute/api_scalar.h:289:20: note: declared here
  289 |   std::vector<int> indices;
      |                    ^~~~~~~


Is it better to add the deprecation on the constructors which use `indices`? I initially didn't think so as users can still subsequently be using `StructFieldOptions::indices` without any deprecation notice, and seems anyway constructing from indices ought to be alright.

File: cpp/src/arrow/compute/kernels/scalar_nested.cc (hunk 11 above)
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested.cc]: Indeed. Think I was just on a bender for dealing with structs. This way works and is easier on the eyes.

File: cpp/src/arrow/compute/api_scalar.h (hunk 9 above)
Comment (milesgranger) [reply cpp/src/arrow/compute/api_scalar.h:290]: ping @pitrou :)

Comment (pitrou) [reply cpp/src/arrow/compute/api_scalar.h:290]: Sorry @milesgranger . I think we should simply remove this member. But we should keep the corresponding constructor.

Comment (pitrou) [reply cpp/src/arrow/compute/api_scalar.h:290]: Btw feel free to ask if you need more guidance/help on this :-)

Comment (milesgranger) [reply cpp/src/arrow/compute/api_scalar.h:290]: Great thanks! I'll take you up on that offer right away. :wink: 

When removing `indices`, I get a compilation error for missing `operator<<` since `FieldRef` is now used in the `kStructFieldOptionsType`'s `DataMember` in `api_scalar.cc`

<details>
<summary>snippet of error</summary>
<\br>
<code>
/home/milesg/Projects/arrow/cpp/src/arrow/compute/function_internal.h:663:3:   required from 'const arrow::compute::FunctionOptionsType* arrow::compute::internal::GetFunctionOptionsType(const Properties& ...) [with Options = arrow::compute::StructFieldOptions; Properties = {arrow::internal::DataMemberProperty<arrow::compute::StructFieldOptions, arrow::FieldRef>}]'
/home/milesg/Projects/arrow/cpp/src/arrow/compute/api_scalar.cc:362:60:   required from here
/home/milesg/Projects/arrow/cpp/src/arrow/compute/function_internal.h:102:6: error: no match for 'operator<<' (operand types are 'std::stringstream' {aka 'std::__cxx11::basic_stringstream<char>'} and 'const arrow::FieldRef')
  102 |   ss << value;
      |   ~~~^~~~~~~~
</code>
</details>

which is reminiscent of [a similar error I'm having in another PR](https://github.com/apache/arrow/pull/14395#discussion_r1012926565)

I _think_ I know how to go about fixing it, implementing a similar `GenericToString` in `function_internal.h`, but as with that PR, not sure where to put that `operator<<` impl.

File: python/pyarrow/_compute.pyx [hunk 12]
```diff
@@ -1322,7 +1322,39 @@ class MakeStructOptions(_MakeStructOptions):
...
+        cdef:
+            CFieldRef field_ref
+            const CFieldRef* field_ref_ptr
+
+        # List[str]/List[bytes] converted to '.a.dotted.path'
+        if isinstance(indices, list) and len(indices):
+            if isinstance(indices[0], str):
+                indices = '.' + '.'.join(indices)
+            elif isinstance(indices[0], bytes):
+                indices = b'.' + b'.'.join(indices)
```
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:1334]: This disallows doing a mix of string and int? Alternatively, I would maybe avoid converting a list to a DottedPath, but rather add a generic list handling case. 

More in general, I think we should align this more with the existing method to create a FieldRef in Python/cython (eg `FieldRef._nested_field` constructor). If we have a list input here, we can just pass it through, and then handle the resulting Expression as you already do here.

File: cpp/src/arrow/compute/kernels/scalar_nested_test.cc [hunk 13]
```diff
@@ -141,16 +148,25 @@ TEST(TestScalarNested, StructField) {
...
+                &extract0_field_ref_name);
+    CheckScalar("struct_field", {arr}, ArrayFromJSON(int64(), "[10, 11, 12, null]"),
+                &extract20_field_ref_nest);
+
     EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid,
                                     ::testing::HasSubstr("out-of-bounds field reference"),
                                     CallFunction("struct_field", {arr}, &invalid1));
     EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid,
-                                    ::testing::HasSubstr("out-of-bounds field reference"),
+                                    ::testing::HasSubstr("No match for FieldRef"),
```
[the remark this answers is not in the document]
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:163]: Ya sorry, this was just a sub-string to the whole error, which is:
> No match for FieldRef.FieldPath(2 4) in struct<a: int32, b: string, c: struct<d: int64, e: double>>

Is this okay, and/or should I include the whole/more of it into the `HasSubstr`check?

File: python/pyarrow/_compute.pyx (hunk 12 above)
Comment (milesgranger) [reply python/pyarrow/_compute.pyx]: That's good! Updated in https://github.com/apache/arrow/pull/14495/commits/b3219ff46563a22af2e7449569e6172ba70c875c

File: cpp/src/arrow/compute/kernels/scalar_nested_test.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:162]: My comment wasn't really about only checking a substring. But that the former error message (which included "out-of-bounds field reference") seems more informative as "No match for FieldRef.FieldPath(2 4) in ...")

So my question is not to update the HasSubtr test line here, but rather: is it possible / would it be worth to try to update `FieldRef::FindOne` (where the error comes from now, I think) to also give such a more informative error message (so we don't loose this more informative error message)

(this might not necessarily be a straightforward fix, didn't check how `FindOne` currently does error checking, so even if we would like to update this, it could also be for a follow-up JIRA)

File: python/pyarrow/_compute.pyx (hunk 12 above)
Comment (jorisvandenbossche) [reply python/pyarrow/_compute.pyx]: Thanks! Can you also add a test for the changed `pc.field(..)` behaviour itself? (without using the struct_field kernel)

File: cpp/src/arrow/engine/substrait/expression_internal.cc [hunk 14]
```diff
@@ -170,9 +170,10 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr,
               out = compute::field_ref(FieldRef(*out_ref, index));
             } else if (out->call() && out->call()->function_name == "struct_field") {
               // Nested StructFields on top of an arbitrary expression
-              std::static_pointer_cast<arrow::compute::StructFieldOptions>(
-                  out->call()->options)
-                  ->indices.push_back(index);
+              auto* field_options =
+                  checked_cast<compute::StructFieldOptions*>(out->call()->options.get());
+              field_options->field_ref =
+                  FieldRef(std::move(field_options->field_ref), index);
```
Comment (pitrou) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:176]: @westonpace @vibhatha Can you look at the Substrait changes here?

File: cpp/src/arrow/engine/substrait/expression_internal.cc [hunk 15]
```diff
@@ -1019,13 +1020,16 @@ Result<std::unique_ptr<substrait::Expression>> ToProto(
 
   if (call->function_name == "struct_field") {
     // catch the special case of calls convertible to a StructField
+    const auto& field_options =
+        checked_cast<const compute::StructFieldOptions&>(*call->options);
+    const DataType& struct_type = *call->arguments[0].type();
+    DCHECK_EQ(struct_type.id(), Type::STRUCT);
```
Comment (pitrou) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:1026]: (note this might be wrong if the argument can be a union?)

File: cpp/src/arrow/compute/kernels/scalar_nested_test.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:163]: Hmm, the latter message actually seems more informative to me, if it includes the entire FieldRed and struct type string representation. But I don't know exactly how the former error message looked :-)

[the remark this answers is not in the document]
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:163]: The former error message was like this: 
`out-of-bounds field reference to field 4 in type struct<a: int32, b: string, c: struct<d: int64, e: double>> with 3 fields`

@jorisvandenbossche Should have made it clearer that I did understand your comment, I only wanted to confirm you were aware the 'new' error did contain additional context and if that was as good as the previous. I'll look more deeply at how to retain the previous error if that's preferable, I don't have much of an opinion on this.

File: python/pyarrow/_compute.pyx (hunk 12 above)
Comment (milesgranger) [reply python/pyarrow/_compute.pyx]: https://github.com/apache/arrow/pull/14495/commits/8fbe3380f546d33273d17a73e2bc133d09b6f38b

File: cpp/src/arrow/compute/kernels/scalar_nested_test.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:163]: Looked around and tried a bit, but the point at which the context of the error is being created in `FieldRef::FindAll(const FieldVector& fields) const` no longer has the context of the `DataType` and thus cannot directly discover `type.num_fields()` as `CheckIndex` does. May be possible return the `out_of_range_depth` variable there back out to the call `FindOne(DataType)` in a error status.

Anyhow, I'd happy to open a JIRA and take it as a follow up if that's desired. :+1: 

A summary of where the new error is coming from:
```cpp
  FieldPath field_path;
  if (field_ref.IsNested() || field_ref.IsName()) {
    // New error happens here in FindOne
    ARROW_ASSIGN_OR_RAISE(field_path, field_ref.FindOne(*type));
  } else {
    field_path = *field_ref.field_path();
  }

  for (const auto& index : field_path.indices()) {
    // Ideally, we get the error from here, CheckIndex
    RETURN_NOT_OK(StructFieldFunctor::CheckIndex(index, *type));
    type = type->field(index)->type().get();
  }
```

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:163]: > Anyhow, I'd happy to open a JIRA and take it as a follow up if that's desired. +1

Yes, let's file a JIRA for this

[the remark this answers is not in the document]
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:163]: [ARROW-18295](https://issues.apache.org/jira/browse/ARROW-18295) :ok_hand:

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 14 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:176]: @pitrou Sure will take a look.

Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:176]: @westonpace does the string fields matter here?

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 15 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:1026]: It might be possible to assume here that `field_ref` consists of a FieldPath with indices already, and then the FindOne and struct type are not needed, and this should also work for a union

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 14 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:176]: in FromProto, I assume the indices will always be integers (since that's how substrait's field references work?)

File: cpp/src/arrow/compute/kernels/scalar_nested_test.cc [hunk 16]
```diff
@@ -141,16 +148,25 @@ TEST(TestScalarNested, StructField) {
...
                                     CallFunction("struct_field", {arr}, &invalid2));
     EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid,
                                     ::testing::HasSubstr("out-of-bounds field reference"),
                                     CallFunction("struct_field", {arr}, &invalid3));
-    EXPECT_RAISES_WITH_MESSAGE_THAT(TypeError, ::testing::HasSubstr("cannot subscript"),
+    EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid,
+                                    ::testing::HasSubstr("No match for FieldRef"),
```
[the remark this answers is not in the document]
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:169]: [ARROW-18295](https://issues.apache.org/jira/browse/ARROW-18295)

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 14 above)
Comment (pitrou) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:176]: @jorisvandenbossche Ah, you might be right.

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 15 above)
Comment (pitrou) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:1026]: I think those are orthogonal concerns: whether this can be a union, and whether the field_ref can be something else than a list of indices.

File: cpp/src/arrow/compute/kernels/scalar_nested_test.cc [hunk 17]
```diff
@@ -141,16 +148,25 @@ TEST(TestScalarNested, StructField) {
...
     CheckScalar("struct_field", {arr}, ArrayFromJSON(int64(), "[10, 11, 12, null]"),
                 &extract20);
+
+    CheckScalar("struct_field", {arr}, ArrayFromJSON(int32(), "[1, null, 3, null]"),
+                &extract0_field_ref_path);
+    CheckScalar("struct_field", {arr}, ArrayFromJSON(int32(), "[1, null, 3, null]"),
+                &extract0_field_ref_name);
+    CheckScalar("struct_field", {arr}, ArrayFromJSON(int64(), "[10, 11, 12, null]"),
+                &extract20_field_ref_nest);
+
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_nested_test.cc:158]: Can you also test these field refs in the union cases below?

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 14 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:176]: Yes that's what I am referring to, Substrait doesn't consider names, only integers (position basically). 
But want to see @westonpace thoughts on this too.

File: cpp/src/arrow/compute/kernels/scalar_nested_test.cc (hunk 17 above)
Comment (milesgranger) [reply cpp/src/arrow/compute/kernels/scalar_nested_test.cc:158]: Updated in https://github.com/apache/arrow/pull/14495/commits/c5675e87fe87770db5b8b1a4989f3ce17f4ae50c

File: python/pyarrow/tests/test_compute.py [hunk 18]
```diff
@@ -2692,14 +2692,32 @@ def test_struct_fields_options():
...
+    assert pc.struct_field(arr, pc.field(('a',))) == a
+
+    assert pc.struct_field(arr, indices=[1, 1]) == b
+    assert pc.struct_field(arr, (1, 1)) == b
+    assert pc.struct_field(arr, [0]) == a
     assert pc.struct_field(arr, []) == arr
 
-    with pytest.raises(TypeError, match="an integer is required"):
-        pc.struct_field(arr, indices=['a'])
+    with pytest.raises(pa.ArrowInvalid, match="No match for FieldRef*"):
```
Comment (pitrou) [inline python/pyarrow/tests/test_compute.py:2712]: Note that `match` is a regular expression, so `*` does not work like that.

Comment (milesgranger) [reply python/pyarrow/tests/test_compute.py]: ah, thought it was glob pattern for some reason, thanks. https://github.com/apache/arrow/pull/14495/commits/ec4023b551448ed6aa0c0ccccd4311c7a1a87618

**Comment by pitrou** [comment]
@jorisvandenbossche Can you give this a final review?

**Comment by jorisvandenbossche** [comment]
Yes, it looks good to me, mostly waiting on the substrait comments.

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 15 above)
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:1026]: The `ToProto` path is very much "for debugging & testing purposes only" so I don't know that we need to consider too much the union case (e.g. the `DCHECK` is sufficient).  That being said, I opened https://github.com/substrait-io/substrait/issues/369 as it is an interesting idea to maybe support something like this in Substrait, which does have unions.

If we want to broaden `ToProto` for use outside of unit tests then we'd want to return an invalid status instead of a `DCHECK` because users can make whatever crazy invalid call arguments they want (e.g. they could have multiple arguments as well which isn't checked here).

We don't need to worry about integers vs. names though.  The field ref is provided to `FindOne` which should translate names into a field path (with indices).  So this should work even if the user supplies a field ref with names.

In summary, +1 for the current implementation.

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 14 above)
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:176]: Yes, `FromProto` will always have integer indices (note that `index` comes from `ref->struct_field().field()` which is `int`).  So I don't see any harm here.  This looks good to me.

**Review by jorisvandenbossche — APPROVED** [review]

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 15 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:1026]: Thanks for the input @westonpace !

**Comment by milesgranger** [comment]
Ping for another review / comments. :) 
(Current failures seem unrelated)

**Comment by jorisvandenbossche** [comment]
>  (Current failures seem unrelated)

Yes, they are also happening on master it seems
